File: haskell/rotational-cipher/src/RotationalCipher.hs
module RotationalCipher (rotate) where

import Data.List (elemIndex, length)
import Data.Char (isLower, isLetter)
import Data.Maybe (fromJust)

lowerCaseLetters :: [Char]
lowerCaseLetters = ['a' .. 'z']

upperCaseLetters :: [Char]
upperCaseLetters = ['A' .. 'Z']

keyForLetter :: Char -> [Char]
keyForLetter letter = if isLower letter then lowerCaseLetters else upperCaseLetters

rotateLetterUsingKey :: Int -> [Char] -> Char -> Char
rotateLetterUsingKey shiftKey key letter = key !! (shiftIndex `mod` length key)
    where
        shiftIndex = shiftKey + (fromJust $ elemIndex letter key)

rotateLetter :: Int -> Char -> Char
rotateLetter shiftKey letter 
    | isLetter letter = rotateLetterUsingKey shiftKey (keyForLetter letter) letter
    | otherwise = letter

rotate :: Int -> String -> String
rotate shiftKey text = map (rotateLetter shiftKey) text

File: scala/rotational-cipher/src/main/scala/RotationalCipher.scala
object RotationalCipher {
  private val LOWERCASE_LETTERS = "abcdefghijklmnopqrstuvwxyz"
  private val UPPERCASE_LETTERS = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"

  def rotate(text: String, shiftKey: Int): String =
    text.map(rotateLetter(shiftKey, _)).mkString

  private def keyForLetter(letter: Char) =
    if (letter.isLower) LOWERCASE_LETTERS else UPPERCASE_LETTERS

  private def rotateLetterUsingKey(shiftKey: Int, letter: Char, key: String): Char = {
    val shiftIndex = key.indexOf(letter) + shiftKey
    key(shiftIndex % key.length)
  }

  private def rotateLetter(shiftKey: Int, char: Char) =
    if (char.isLetter)
      rotateLetterUsingKey(shiftKey, char, keyForLetter(char))
    else
      char
}
